### PR TITLE
refactor: implement native periodic thread

### PR DESCRIPTION
**Pull request by P403n1x87**
This PR re-implements the periodic service to use a periodic thread implemented in C++. The reason for doing this is to reduce the dependency of the library on the `threading` module, which is the source of issues when it comes to frameworks such as `gevent`.

This is the first step towards a refactoring that is aimed at removing the dependency on the `threading` module entirely, potentially leading to phasing out the module cloning mechanism, which is what we currently have in place to support frameworks such as the already mentioned `gevent`.

## Testing strategy

The existing test suite provides already many reliable test cases that can give us good confidence about the proposed changes. Further testing is being performed in internal environments.

## Performance

We do not expect any particular performance impact from this change

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: ddtrace/commands/ddtrace_run.py [hunk 1]
```diff
@@ -147,3 +147,7 @@ def main():
         raise
 
     sys.exit(0)
+
+
+if __name__ == "__main__":
```
Comment (P403n1x87) [inline ddtrace/commands/ddtrace_run.py:152]: This makes it easy to use lldb with ddtrace-run

File: tests/conftest.py
Comment (P403n1x87) [inline tests/conftest.py:1]: This can be taken out in its own PR

File: ddtrace/commands/ddtrace_run.py (hunk 1 above)
Comment (emmettbutler) [reply ddtrace/commands/ddtrace_run.py:147]: Nice

File: ddtrace/internal/_threads.cpp [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+static int
+PeriodicThread_init(PeriodicThread* self, PyObject* args, PyObject* kwargs)
+{
+    static const char* kwlist[] = { "interval", "target", "name", "on_shutdown", NULL };
+
+    self->name = Py_None;
+    self->_on_shutdown = Py_None;
+
+    if (!PyArg_ParseTupleAndKeywords(
+          args, kwargs, "dO|OO", (char**)kwlist, &self->interval, &self->_target, &self->name, &self->_on_shutdown))
```
Comment (emmettbutler) [inline ddtrace/internal/_threads.cpp:165]: Is this indentation and bracketing style enforced by a linter?

**Review by emmettbutler — APPROVED** [review]

**Comment by P403n1x87** [comment]
> I wonder why the benchmarking comment hasn't appeared on this pull request. I'm interested in the performance implications.

Unfortunately they won't run on my fork 😞

File: ddtrace/internal/_threads.cpp [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+        GILGuard _;
+        Py_DECREF(_obj);
+    }
+
+  private:
+    PyObject* _obj;
+};
+
+// ----------------------------------------------------------------------------
+class Event
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/internal/_threads.cpp:79]: Yep this should essentially be equivalent to its Python counterpart 🤞

File: tests/opentracer/test_tracer_gevent.py
Comment (P403n1x87) [inline tests/opentracer/test_tracer_gevent.py:1]: The formatter demanded this change 🤷

File: .circleci/config.templ.yml [hunk 4]
```diff
@@ -37,7 +37,7 @@ machine_executor: &machine_executor
     # https://support.circleci.com/hc/en-us/articles/360045268074-Build-Fails-with-Too-long-with-no-output-exceeded-10m0s-context-deadline-exceeded-
     - PYTHONUNBUFFERED: 1
   steps:
-    - run: sudo apt-get update && sudo DEBIAN_FRONTEND=noninteractive apt-get install -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" --force-yes -y --no-install-recommends openssh-client git
+    - run: sudo apt-get update && sudo DEBIAN_FRONTEND=noninteractive apt-get install -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" --force-yes -y --no-install-recommends openssh-client git gdb
```
Comment (P403n1x87) [inline .circleci/config.templ.yml:40]:
```suggestion
    - run: sudo apt-get update && sudo DEBIAN_FRONTEND=noninteractive apt-get install -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" --force-yes -y --no-install-recommends openssh-client git
```

File: ddtrace/internal/_threads.cpp [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+    std::mutex _mutex;
+    bool _set = false;
+};
+
+// ----------------------------------------------------------------------------
+typedef struct periodic_thread
+{
+    PyObject_HEAD
+
+      double interval;
```
Comment (brettlangdon) [inline ddtrace/internal/_threads.cpp:113]:
```suggestion
    double interval;
```

File: ddtrace/internal/_threads.pyi [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    def __init__(
+        self,
+        interval: float,
+        target: t.Callable,
+        name: t.Optional[str] = None,
+        on_shutdown: t.Optional[t.Callable] = None,
+    ) -> None: ...
+    def start(self) -> None: ...
+    def stop(self) -> None: ...
+    def join(self, timeout: t.Optional[float] = None) -> None: ...
```
Comment (brettlangdon) [inline ddtrace/internal/_threads.pyi:16]: missing awake?

File: ddtrace/profiling/collector/stack.pyx [hunk 7]
```diff
@@ -292,7 +294,7 @@ cdef stack_collect(ignore_profiler, thread_time, max_nframes, interval, wall_tim
     # Do not use `threading.enumerate` to not mess with locking (gevent!)
     thread_id_ignore_list = {
         thread_id
-        for thread_id, thread in ddtrace_threading._active.items()
+        for thread_id, thread in chain(periodic_threads.items(), ddtrace_threading._active.items())
```
Comment (brettlangdon) [inline ddtrace/profiling/collector/stack.pyx:297]: are these threads even visible to Python?

**Review by brettlangdon — COMMENTED** [review]
minor nits/questions, otherwise lgtm

File: ddtrace/profiling/collector/stack.pyx (hunk 7 above)
Comment (P403n1x87) [reply ddtrace/profiling/collector/stack.pyx:297]: They are not accessible from Python, but when you traverse the list of thread states from the runtime state you will hit them. The profiler currently profiles itself, so not doing this would be a "breaking change" i suppose

File: ddtrace/internal/_threads.pyi (hunk 6 above)
Comment (P403n1x87) [reply ddtrace/internal/_threads.pyi:16]: Ah yes, thank you!

File: ddtrace/internal/_threads.cpp (hunk 5 above)
Comment (P403n1x87) [reply ddtrace/internal/_threads.cpp:113]: 🤔 I don't know if the formatter would bark at us for this. This is how my IDE has decided to format this line 🤷

File: ddtrace/profiling/collector/stack.pyx (hunk 7 above)
Comment (brettlangdon) [reply ddtrace/profiling/collector/stack.pyx:297]: > They are not accessible from Python, but when you traverse the list of thread states from the runtime state you will hit them.

This is good context that might be worth capturing in a comment somewhere.

File: ddtrace/internal/_threads.cpp (hunk 5 above)
Comment (brettlangdon) [reply ddtrace/internal/_threads.cpp:113]: oh, maybe because the `PyObject_HEAD` above with no `;`? might have it confused?

**Comment by brettlangdon** [comment]
Running flask-simple locally:

| Benchmark               | ddtrace==2.8.3 | native periodic thread |
|-------------------------|----------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------|
| flasksimple-profiler    | 701 us                                                                           | 686 us: 1.02x faster                                                                                |
| flasksimple-debugger    | 595 us                                                                           | 557 us: 1.07x faster                                                                                |
| flasksimple-iast-get    | 560 us                                                                           | 550 us: 1.02x faster                                                                                |
| flasksimple-appsec-get  | 1.21 ms                                                                          | 1.13 ms: 1.07x faster                                                                               |
| flasksimple-appsec-post | 1.01 ms                                                                          | 1.03 ms: 1.02x slower                                                                               |
| Geometric mean          | (ref)                                                                            | 1.02x faster                                                                                        |

Benchmark hidden because not significant (3): flasksimple-baseline, flasksimple-tracer, flasksimple-appsec-telemetry

**Review by emmettbutler — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

**Review by gnufede — APPROVED** [review]

**Comment by emmettbutler** [comment]
@P403n1x87 this seems to be a proximate cause of failures like [this](https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/64360/workflows/e632b63f-3e1e-468a-be74-7c9476e40670/jobs/3982041) on `main`.
